File: crates/python/mirascript/_vm/lib/vm_global/mod/complex/math.py
"""Transcendental Cartesian complex operations."""

from __future__ import annotations

import math

from .basic import multiply, polar
from .utils import C, INF, NAN, div, math_call, maximum, minimum, radial


def _log_abs(z: C) -> float:
    x, y = z
    a, b = maximum(abs(x), abs(y)), minimum(abs(x), abs(y))
    if a == INF:
        return INF
    if a == 0:
        return -INF
    return math_call(math.log, a) + 0.5 * math_call(math.log1p, div(b, a) * div(b, a))


def log(z: C) -> C:
    return (_log_abs(z), math.atan2(z[1], z[0]))


def log_base(z: C, base: float) -> C:
    a, b = log(z)
    return (a / base, b / base)


def sqrt(z: C) -> C:
    x, y = z
    if abs(y) == INF:
        return (INF, y)
    if x == INF:
        return (INF, NAN if math.isnan(y) else math.copysign(0, y))
    if x == -INF:
        return (NAN if math.isnan(y) else 0.0, math.copysign(INF, y))
    s = maximum(abs(x), abs(y))
    if s == 0:
        return (0.0, y)
    t = math_call(math.sqrt, s) * math_call(
        math.sqrt, (math.hypot(x / s, y / s) + abs(x) / s) / 2
    )
    return (t, div(y, t) / 2) if x >= 0 else (div(abs(y), t) / 2, math.copysign(t, y))


def exp(z: C) -> C:
    return polar(math_call(math.exp, z[0]), z[1])


def expm1(z: C) -> C:
    x, y = z
    half_sin = math_call(math.sin, y / 2)
    return (
        math_call(math.expm1, x) * math_call(math.cos, y) - 2 * half_sin * half_sin,
        radial(math_call(math.exp, x), math_call(math.sin, y)),
    )


def log1p(z: C) -> C:
    x, y = z
    re = (
        0.5 * math_call(math.log1p, x * (2 + x) + y * y)
        if abs(x) < 0.5 and abs(y) < 0.5
        else _log_abs((1 + x, y))
    )
    return (re, math.atan2(y, 1 + x))


def pow(z: C, w: C) -> C:
    if w[0] == 0 and w[1] == 0:
        return (1.0, 0.0)
    if w[0] == 1 and w[1] == 0:
        return z
    if z[0] == 0 and z[1] == 0:
        if w[1] != 0 or math.isnan(w[0]):
            return (NAN, NAN)
        return polar(0.0 if w[0] > 0 else INF, w[0] * math.atan2(z[1], z[0]))
    if w[1] == 0:
        if z[1] == 0 and math.isfinite(w[0]) and w[0].is_integer():
            try:
                re = math.pow(z[0], w[0])
            except OverflowError:
                re = -INF if z[0] < 0 and w[0] % 2 != 0 else INF
            except ValueError:
                re = NAN
            return (re, math.copysign(0, z[1]))
        return polar(
            math_call(math.exp, _log_abs(z) * w[0]),
            math.atan2(z[1], z[0]) * w[0],
        )
    return exp(multiply(w, log(z)))


def cbrt(z: C) -> C:
    return polar(math_call(math.exp, _log_abs(z) / 3), math.atan2(z[1], z[0]) / 3)


def sin(z: C) -> C:
    x, y = z
    return (
        radial(math_call(math.cosh, y), math_call(math.sin, x)),
        radial(math_call(math.sinh, y), math_call(math.cos, x)),
    )


def cos(z: C) -> C:
    x, y = z
    return (
        radial(math_call(math.cosh, y), math_call(math.cos, x)),
        -radial(math_call(math.sinh, y), math_call(math.sin, x)),
    )


def sinh(z: C) -> C:
    x, y = z
    return (
        radial(math_call(math.sinh, x), math_call(math.cos, y)),
        radial(math_call(math.cosh, x), math_call(math.sin, y)),
    )


def cosh(z: C) -> C:
    x, y = z
    return (
        radial(math_call(math.cosh, x), math_call(math.cos, y)),
        radial(math_call(math.sinh, x), math_call(math.sin, y)),
    )


def tan(z: C) -> C:
    x, y = z
    t = math_call(math.exp, -2 * abs(y))
    cx, sx = math_call(math.cos, x), math_call(math.sin, x)
    denominator = (1 - t) * (1 - t) + 4 * t * cx * cx
    return (
        div(4 * t * sx * cx, denominator),
        math.copysign(div(-math_call(math.expm1, -4 * abs(y)), denominator), y),
    )


def tanh(z: C) -> C:
    a, b = tan((z[1], z[0]))
    return (b, a)


def asin(z: C) -> C:
    x, y = z
    if y == 0 and abs(x) <= 1:
        return (math.asin(x), y)
    s = maximum(abs(x), abs(y))
    if s > 1e150:
        return (math.atan2(x, abs(y)), math.copysign(_log_abs(z) + math.log(2), y))
    h, im = _inverse_geometry(x, y)
    return (math.atan2(x, h), im)


def _inverse_geometry(x: float, y: float) -> C:
    ax = abs(x)
    r, t = math.hypot(ax + 1, y), math.hypot(ax - 1, y)
    a = r / 2 + t / 2
    delta = (
        math.hypot(
            div(y, math_call(math.sqrt, r + (1 + ax))),
            div(y, math_call(math.sqrt, t + (1 - ax))),
        )
        / math.sqrt(2)
        if ax <= 1
        else math_call(math.sqrt, a - 1)
    )
    gap = (
        math.hypot(math_call(math.sqrt, 1 - ax), delta)
        if ax <= 1
        else math.hypot(
            div(y, math_call(math.sqrt, r + (ax + 1))),
            div(y, math_call(math.sqrt, t + (ax - 1))),
        )
        / math.sqrt(2)
    )
    return (
        gap * math_call(math.sqrt, a + ax),
        math.copysign(
            math_call(math.log1p, delta * (delta + math_call(math.sqrt, a + 1))),
            y,
        ),
    )


def acos(z: C) -> C:
    x, y = z
    if y == 0 and abs(x) <= 1:
        return (math.acos(x), -y)
    if maximum(abs(x), abs(y)) > 1e150:
        return (
            math.atan2(abs(y), x),
            -math.copysign(_log_abs(z) + math.log(2), y),
        )
    h, im = _inverse_geometry(x, y)
    return (math.atan2(h, x), -im)


def atan(z: C) -> C:
    x, y = z
    if y == 0:
        return (math.atan(x), y)
    re = (math.atan2(x, 1 - y) + math.atan2(x, 1 + y)) / 2
    h = math.hypot(x, y - 1)
    if h == 0:
        im = INF
    elif abs(y) < 0.5 * h:
        im = 0.25 * math_call(math.log1p, div(div(4 * y, h), h))
    else:
        im = (_log_abs((x, y + 1)) - _log_abs((x, y - 1))) / 2
    return (re, im)


def asinh(z: C) -> C:
    a, b = asin((-z[1], z[0]))
    return (b, -a)


def acosh(z: C) -> C:
    a, b = acos(z)
    return (abs(b), math.copysign(a, z[1]))


def atanh(z: C) -> C:
    a, b = atan((-z[1], z[0]))
    return (b, -a)

File: crates/vm/src/standard_library/module/complex/math.rs
use std::f64::consts::LN_2;

use super::{C, INF, NAN, multiply, polar, radial};

fn log_abs((x, y): C) -> f64 {
    if x.is_nan() || y.is_nan() {
        return NAN;
    }
    let a = x.abs().max(y.abs());
    let b = x.abs().min(y.abs());
    if a == INF {
        return INF;
    }
    if a == 0.0 {
        return -INF;
    }
    a.ln() + 0.5 * ((b / a).powi(2)).ln_1p()
}
pub(super) fn log(z: C) -> C {
    (log_abs(z), z.1.atan2(z.0))
}
pub(super) fn log_base(z: C, base: f64) -> C {
    let (a, b) = log(z);
    (a / base, b / base)
}
pub(super) fn sqrt((x, y): C) -> C {
    if y.abs() == INF {
        return (INF, y);
    }
    if x == INF {
        return (INF, if y.is_nan() { NAN } else { 0.0_f64.copysign(y) });
    }
    if x == -INF {
        return (if y.is_nan() { NAN } else { 0.0 }, INF.copysign(y));
    }
    if x.is_nan() || y.is_nan() {
        return (NAN, NAN);
    }
    let s = x.abs().max(y.abs());
    if s == 0.0 {
        return (0.0, y);
    }
    let t = s.sqrt() * (((x / s).hypot(y / s) + x.abs() / s) / 2.0).sqrt();
    if x >= 0.0 {
        (t, y / t / 2.0)
    } else {
        (y.abs() / t / 2.0, t.copysign(y))
    }
}
pub(super) fn exp((x, y): C) -> C {
    polar(x.exp(), y)
}
pub(super) fn expm1((x, y): C) -> C {
    (
        x.exp_m1() * y.cos() - 2.0 * (y / 2.0).sin().powi(2),
        radial(x.exp(), y.sin()),
    )
}
pub(super) fn log1p((x, y): C) -> C {
    let re = if x.abs() < 0.5 && y.abs() < 0.5 {
        0.5 * (x * (2.0 + x) + y * y).ln_1p()
    } else {
        log_abs((1.0 + x, y))
    };
    (re, y.atan2(1.0 + x))
}
pub(super) fn pow(z: C, w: C) -> C {
    if w.0 == 0.0 && w.1 == 0.0 {
        return (1.0, 0.0);
    }
    if w.0 == 1.0 && w.1 == 0.0 {
        return z;
    }
    if z.0 == 0.0 && z.1 == 0.0 {
        if w.1 != 0.0 || w.0.is_nan() {
            return (NAN, NAN);
        }
        return polar(if w.0 > 0.0 { 0.0 } else { INF }, w.0 * z.1.atan2(z.0));
    }
    if w.1 == 0.0 {
        if z.1 == 0.0 && w.0.is_finite() && w.0.trunc() == w.0 {
            return (z.0.powf(w.0), 0.0_f64.copysign(z.1));
        }
        return polar((log_abs(z) * w.0).exp(), z.1.atan2(z.0) * w.0);
    }
    exp(multiply(w, log(z)))
}
pub(super) fn cbrt(z: C) -> C {
    polar((log_abs(z) / 3.0).exp(), z.1.atan2(z.0) / 3.0)
}
pub(super) fn sin((x, y): C) -> C {
    (radial(y.cosh(), x.sin()), radial(y.sinh(), x.cos()))
}
pub(super) fn cos((x, y): C) -> C {
    (radial(y.cosh(), x.cos()), -radial(y.sinh(), x.sin()))
}
pub(super) fn sinh((x, y): C) -> C {
    (radial(x.sinh(), y.cos()), radial(x.cosh(), y.sin()))
}
pub(super) fn cosh((x, y): C) -> C {
    (radial(x.cosh(), y.cos()), radial(x.sinh(), y.sin()))
}
pub(super) fn tan((x, y): C) -> C {
    let t = (-2.0 * y.abs()).exp();
    let denominator = (1.0 - t).powi(2) + 4.0 * t * x.cos().powi(2);
    (
        4.0 * t * x.sin() * x.cos() / denominator,
        (-(-4.0 * y.abs()).exp_m1() / denominator).copysign(y),
    )
}
pub(super) fn tanh((x, y): C) -> C {
    let (a, b) = tan((y, x));
    (b, a)
}
pub(super) fn asin((x, y): C) -> C {
    if y == 0.0 && x.abs() <= 1.0 {
        return (x.asin(), y);
    }
    let s = if x.is_nan() || y.is_nan() {
        NAN
    } else {
        x.abs().max(y.abs())
    };
    if s > 1e150 {
        return (x.atan2(y.abs()), (log_abs((x, y)) + LN_2).copysign(y));
    }
    let (h, im) = inverse_geometry(x, y);
    (x.atan2(h), im)
}
fn inverse_geometry(x: f64, y: f64) -> C {
    let ax = x.abs();
    let r = (ax + 1.0).hypot(y);
    let t = (ax - 1.0).hypot(y);
    let a = r / 2.0 + t / 2.0;
    let delta = if ax <= 1.0 {
        (y / (r + (1.0 + ax)).sqrt()).hypot(y / (t + (1.0 - ax)).sqrt()) / std::f64::consts::SQRT_2
    } else {
        (a - 1.0).sqrt()
    };
    let gap = if ax <= 1.0 {
        (1.0 - ax).sqrt().hypot(delta)
    } else {
        (y / (r + (ax + 1.0)).sqrt()).hypot(y / (t + (ax - 1.0)).sqrt()) / std::f64::consts::SQRT_2
    };
    (
        gap * (a + ax).sqrt(),
        (delta * (delta + (a + 1.0).sqrt())).ln_1p().copysign(y),
    )
}
pub(super) fn acos((x, y): C) -> C {
    if y == 0.0 && x.abs() <= 1.0 {
        return (x.acos(), -y);
    }
    if !x.is_nan() && !y.is_nan() && x.abs().max(y.abs()) > 1e150 {
        return (y.abs().atan2(x), -(log_abs((x, y)) + LN_2).copysign(y));
    }
    let (h, im) = inverse_geometry(x, y);
    (h.atan2(x), -im)
}
pub(super) fn atan((x, y): C) -> C {
    if y == 0.0 {
        return (x.atan(), y);
    }
    let re = (x.atan2(1.0 - y) + x.atan2(1.0 + y)) / 2.0;
    let h = x.hypot(y - 1.0);
    let im = if h == 0.0 {
        INF
    } else if y.abs() < 0.5 * h {
        0.25 * ((4.0 * y / h) / h).ln_1p()
    } else {
        (log_abs((x, y + 1.0)) - log_abs((x, y - 1.0))) / 2.0
    };
    (re, im)
}
pub(super) fn asinh((x, y): C) -> C {
    let (a, b) = asin((-y, x));
    (b, -a)
}
pub(super) fn acosh(z: C) -> C {
    let (a, b) = acos(z);
    (b.abs(), a.copysign(z.1))
}
pub(super) fn atanh((x, y): C) -> C {
    let (a, b) = atan((-y, x));
    (b, -a)
}
